### ISSUE TITLE
w/ memory-map on and w/o subsampling, how large RAM do I need to run DLRM on the full-sized Terabyte dataset?

### ISSUE THREAD

**Issue by 12eadp001** [facebookresearch/dlrm#199]
I want the full-sized Terabyte dataset, and I learned from one of the previous issues that to preprocess it, we need about 3-4 TB RAM, which is unachievable to me. So I am wondering is there any other method to run this largest setting?

**Comment by mnaumovfb** [comment]
You can try using --memory-map flag, it will minimize the required RAM. This dataset will take a long time to pre-process and you will still need significant disk space, though.

**Comment by 12eadp001** [comment]
> You can try using --memory-map flag, it will minimize the required RAM. This will take a long time to pre-process and you will need significant disk space, though.

I have about 3TB disk space (the 1.1 TB dataset size not included) now and I assume it is enough (correct me if I am wrong). But I have two concerns:
1. what is the memory requirement under this setting? Like will 256 GB RAM work?
2. I am doing the preprocessing in a department cluster and using slurm to schedule my tasks, so I may experience some disconnect problem from time to time (e.g. due to the duration limit of Slurm), I wonder if the preprocessing can resume from a checkpoint or something if it was interrupted?

**Comment by mnaumovfb** [comment]
This might not be enough, you probably will need 4TB+ for storing all the intermediate pre-processing files. In the --memory-map approach the pre-processing shuffles data in and out of disk to preserve RAM and stores many intermediate files. If you do not have a larger machine, my advice is to work with a smaller dataset.

The pre-processing is broken into several stages, you can restart the pre-processing on a stage boundary, but if it's interrupted during a particular stage, you will need to manually remove the files for that stage and restart it, otherwise data might be corrupted, e.g. https://github.com/facebookresearch/dlrm/issues/125#issuecomment-676574041

**Comment by 12eadp001** [comment]
Thanks, I will install larger disks as I really want to play with large models. So I assume the RAM could meet the need?

By the way, would you think that it would be a good idea if you (or the MLCommons community) can provide a test set data loader? For people who are interested in inference only, processing the whole dataset is too expensive, and the test dataloader shouldn't be very large I think.

**Comment by mnaumovfb** [comment]
I believe that 256 GB RAM and more disk space should be enough. The MLPerf/MLCommons community might have a pre-preprocessed dataset. You should ask them about it on their forums.

ps. currently the generation of training and test sets are tied together, so it's not possible to just generate the test set.

**Comment by 12eadp001** [comment]
Yes, I understand that we cannot just generate the test set, but if someone can preprocess all the data, and simply store the test parts in a binary form (as only the last day is used, the data volume is relatively small), users can directly load it to use? Is this understanding correct?

**Comment by gnupdev** [comment]
Hi!
I need the whole preprocessed terabyte dataset for training experiments.

How much RAM and Storage do I need to preprocess whole terabyte dataset??

i already read previous issues, but hard to be sure...

**Comment by 12eadp001** [comment]
@gnupdev Hi, as @mnaumovfb just pointed out, 256 GB RAM + more than 4 TB disk should be enough under the memory map settings. I don't know the requirements without this flag, maybe the maintainers can give you the answer.

**Comment by gnupdev** [comment]
Thanks you!

**Comment by 12eadp001** [comment]
@mnaumovfb Hi I am a bit confused about the "boundary" between stage(i) and stage(ii). Now I have all `day_*.npz` files, together with all `day_fea_dict_*.npz`, and `day_fea_count.npz`, `day_day_count.npz`. Should I delete all except the `day_*.npz` files? You said "be careful" about the two small files, but what does this exactly mean? Thanks.

**Comment by mnaumovfb** [comment]
In general, you should not be touching intermediate pre-processing files. This could only be a solution of less resort if something has crashed and you absolutely need to recover it. It is complicated to make this recovery work correctly and I do not guarantee it will work in every case.

If you are interested in dealing with a large dataset, you will need to have uninterrupted access to a machine/node with at least 256GB RAM and 8TB of disk space for a long period of time (weeks). My advice is to try working with smaller dataset before trying the larger one.

**Comment by 12eadp001** [comment]
> In general, you should not be touching intermediate pre-processing files.

Yes, I don't feel like touching them...but the thing is, as you said, the program had crashed due to the time limitation of my cluster manager. I had to recover it.

>  If you are interested in dealing with a large dataset, you will need to have uninterrupted access to a machine/node with at least 256GB RAM and 8TB of disk space for a long period of time (weeks).

I am confused why it suddenly turns to be 8TB...? Were we just saying 4 TB?

Thanks for reminding me to try smaller one first, I already played with the 0.875-subsampled version, it is already time to fight with this Terabyte monster.
